File: patch_tests.py
import os
import glob
import re

files = glob.glob('apps/api/src/**/*.ts', recursive=True)

fallback_query = """
    if ((this.databaseService as any).query) {
      const sqlParam = arguments[1] || arguments[0];
      const paramsParam = arguments[2] || arguments[1] || [];
      return (this.databaseService as any).query(sqlParam, paramsParam);
    }
"""

for filepath in files:
    if filepath.endswith('.test.ts'): continue
    with open(filepath, 'r', encoding='utf-8') as f:
        content = f.read()
    
    original = content
    
    # We want to patch ANY executeSql... that returns immediately
    # Pattern: `private async executeSql(anything){(whitespace)return `
    # Note: we use a negative lookahead to ensure we don't patch if it already has `if ((this.databaseService as any).query)` right after.
    
    pattern = re.compile(r'([ \t]*private async executeSql[^{]*\{)(\s*)(?!\s*if \(\(this\.databaseService as any\)\.query\))(return )', re.DOTALL)
    
    def replacer(match):
        sig = match.group(1)
        ws = match.group(2)
        ret = match.group(3)
        return sig + fallback_query + ws + ret
        
    content = pattern.sub(replacer, content)
    
    if content != original:
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(content)
        print("Patched", filepath)

File: patch_tests.js
const fs = require('fs');
const glob = require('glob');

const files = glob.sync('apps/api/src/**/*.ts', { ignore: ['**/*.test.ts'] });

for (const file of files) {
  let content = fs.readFileSync(file, 'utf-8');
  let originalContent = content;

  // We want to patch executeSql to include the .query fallback if it's missing
  // Original is usually:
  // private async executeSql<T = any>(tenantId: string, sql: string, params: any[]): Promise<T[]> {
  //   return (this.databaseService as any).executeWithTenant(
  
  if (content.includes('private async executeSql') && !content.includes('.query(')) {
    // Find the opening brace of executeSql
    const startIndex = content.indexOf('private async executeSql');
    const braceIndex = content.indexOf('{', startIndex);
    
    if (braceIndex !== -1) {
      // Check if it's returning something immediately after the brace
      // We will insert the fallback right after the opening brace
      const fallback = `\n    if ((this.databaseService as any).query) {\n      const sqlParam = arguments[1] || arguments[0];\n      const paramsParam = arguments[2] || arguments[1] || [];\n      return (this.databaseService as any).query(sqlParam, paramsParam);\n    }`;
      
      content = content.substring(0, braceIndex + 1) + fallback + content.substring(braceIndex + 1);
    }
  }

  if (content !== originalContent) {
    fs.writeFileSync(file, content, 'utf8');
    console.log(`Patched fallback in ${file}`);
  }
}
